### PR TITLE
feat: enhance homepage and add builders grid

### PR DESCRIPTION
**Pull request by gboigwe**
# Enhanced Homepage and Builders Grid

## Changes Made
- Added BuildersGrid component showing registered builders
- Created API route to fetch builder addresses from filesystem
- Connected grid to personal builder pages
- Enhanced homepage dark mode visibility
- Updated styling for better UX

## Implementation Details
- Scans filesystem for builder pages in `/builders` directory
- Displays builder profiles in responsive grid layout
- Maintains consistent styling with existing components

[Add screenshots of homepage and grid in light/dark modes]
### Screenshot for the Checked Builders Count:
image

### Screenshot for the number of registered builders page:
image

## Additional Information
- Still trying to fix some console issues

- [x] I have read the [contributing docs](/scaffold-eth/scaffold-eth-2/blob/main/CONTRIBUTING.md) (if this is your first contribution)
- [x] This is not a duplicate of any [existing pull request](https://github.com/scaffold-eth/scaffold-eth-2/pulls)

## Related Issues

_Closes #4 _

Your ENS/address: 0x5cc8Be96B1C9A68F57a73b5bEa60cF5D890055A1

### CODE REVIEW

**Comment by gboigwe** [comment]
Although not entirely completed as I still need to work on some console errors I found, feel free to make suggestions and or contribute.
And if there is any issue at all please point it out, even if it is the console error issue, please still point it out

File: packages/nextjs/app/builders/0x5cc8Be96B1C9A68F57a73b5bEa60cF5D890055A1/page.tsx
Comment (melanke) [inline packages/nextjs/app/builders/0x5cc8Be96B1C9A68F57a73b5bEa60cF5D890055A1/page.tsx:1]: I think this file should be left out of this PR. It is part of another issue that will be reviewed separately.

File: packages/nextjs/app/page.tsx [hunk 1]
```diff
@@ -2,48 +2,105 @@
...
+
+const BATCH_REGISTRY_ADDRESS = "0x65E42251A2ca0906c640331afd77f45F4682c56C";
+const BATCH_REGISTRY_ABI = deployedContracts[31337].BatchRegistry.abi;
 
 const Home: NextPage = () => {
+  const { data: checkedInCount, isLoading } = useContractRead({
+    address: BATCH_REGISTRY_ADDRESS,
+    abi: BATCH_REGISTRY_ABI,
+    functionName: "checkedInCounter",
+  });
```
Comment (melanke) [inline packages/nextjs/app/page.tsx:18]: I think you should use `useScaffoldReadContract`, it's way easier to use and would not require the contract address and ABI.

File: packages/nextjs/components/batch/BuildersGrid.tsx [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+//     const directories = await readdir(buildersPath, { withFileTypes: true });
+//     return directories
+//       .filter(dirent => dirent.isDirectory())
+//       .map(dirent => dirent.name)
+//       .filter(name => name.startsWith('0x'));
+//   } catch (error) {
+//     console.error('Error reading builders directory:', error);
+//     return [];
+//   }
+// }
```
Comment (melanke) [inline packages/nextjs/components/batch/BuildersGrid.tsx:23]: You chose to create an api route, so this could be removed, right?

File: packages/nextjs/components/batch/BuildersGrid.tsx [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+          address: address,
+        }));
+        setBuilders(buildersList);
+        setIsLoading(false);
+      })
+      .catch(error => {
+        console.error("Error fetching builders:", error);
+        setIsLoading(false);
+      });
+  }, []);
```
Comment (melanke) [inline packages/nextjs/components/batch/BuildersGrid.tsx:45]: This is a matter of preference, but react-query would fit well here.

Comment (melanke) [reply packages/nextjs/components/batch/BuildersGrid.tsx:45]: Also, beside getting the builders by their profile pages, we could get whoever made checkin, even the ones without a profile page. Check this [discussion](https://github.com/BuidlGuidl/batch11.buidlguidl.com/issues/4#issuecomment-2485640200).

**Review by melanke — CHANGES_REQUESTED** [review]
Nice job. Just a few ideas for improvements

File: packages/nextjs/app/builders/0x5cc8Be96B1C9A68F57a73b5bEa60cF5D890055A1/page.tsx
Comment (gboigwe) [reply packages/nextjs/app/builders/0x5cc8Be96B1C9A68F57a73b5bEa60cF5D890055A1/page.tsx:1]: Okay thanks. The reason for adding it is so that it can be tested, else the purpose of the implementation won't be valid.

Aside that, I tried pushing without it but it wasn't committing successfully, I would share the screenshot later on.

There was a warning signthat expected it but it was not found.

Thanks for the observations, I will still sit with it and checkout how I can do just that

File: packages/nextjs/components/batch/BuildersGrid.tsx (hunk 2 above)
Comment (gboigwe) [reply packages/nextjs/components/batch/BuildersGrid.tsx:23]: Yeah, thanks, I will do just that, I forgot that it was still there, I used it to test it before moving it to api route

File: packages/nextjs/app/page.tsx (hunk 1 above)
Comment (gboigwe) [reply packages/nextjs/app/page.tsx:18]: Okay, thanks I will do that. Thank you

File: packages/nextjs/components/batch/BuildersGrid.tsx (hunk 3 above)
Comment (gboigwe) [reply packages/nextjs/components/batch/BuildersGrid.tsx:45]: Can we get the details from the builders that checked-in?
Okay, I will try it out and see, but not sure how to figure it, but I will definitely try all my best. Thanks so much

File: packages/nextjs/app/builders/0x5cc8Be96B1C9A68F57a73b5bEa60cF5D890055A1/page.tsx
Comment (melanke) [reply packages/nextjs/app/builders/0x5cc8Be96B1C9A68F57a73b5bEa60cF5D890055A1/page.tsx:1]: The main branch already have my profile page, so you can use it to test. You might need to rebase tho.

Comment (phipsae) [reply packages/nextjs/app/builders/0x5cc8Be96B1C9A68F57a73b5bEa60cF5D890055A1/page.tsx:1]: @gboigwe Yes, please remove anything related to your personal page from the PR 🙏. 

It might be easiest to start fresh with a new branch. As @melanke mentioned, his page is already live, so you could merge the latest version of main into your repo, to try it out. If you wait for a little bit, we will have more personal pages live.

**Comment by gboigwe** [comment]
Made some changes as suggested by @melanke and @phipsae and I have committed them. Sure about the react-query though yet but any other suggestions

**Comment by phipsae** [comment]
@gboigwe thanks, you def tackle one of the more difficult tasks! 

Pls look into the comments @melanke made and work on them before you ask for a review. 
Pls use `useScaffoldReadContract` instead of `useContractRead` for example. 
Also it shows me that you touched the builders profiles, there shouldn't be any changes. 

Let me know then I'll have a review!

**Comment by gboigwe** [comment]
Okay, thank you.

I actually did used the useScaffoldReadContract

I will check to confirm why that was not seen. Thank you

**Comment by gboigwe** [comment]
Oh and for the builders profile, remember that I used my builders page before for testing but @melanke suggested that I pull his as it was merged already, so I removed mine from the PR and used his

File: packages/nextjs/app/builders/0x2Bc096A12C5b37F035180aDeF70EB2B88351e5B8/_components/SocialLinks.tsx
Comment (phipsae) [inline packages/nextjs/app/builders/0x2Bc096A12C5b37F035180aDeF70EB2B88351e5B8/_components/SocialLinks.tsx:1]: There should be no changes made to this file. Shouldn't be touched since he only focus on the builders overview issue #4.

File: packages/nextjs/app/builders/0x2Bc096A12C5b37F035180aDeF70EB2B88351e5B8/page.tsx
Comment (phipsae) [inline packages/nextjs/app/builders/0x2Bc096A12C5b37F035180aDeF70EB2B88351e5B8/page.tsx:1]: There should be no changes made to this file. Shouldn't be touched since he only focus on the builders overview issue #4.

File: packages/nextjs/contracts/deployedContracts.ts
Comment (phipsae) [inline packages/nextjs/contracts/deployedContracts.ts:1]: There should be no changes made to this file, since We are not adding any new contract.

File: packages/nextjs/next.config.js
Comment (phipsae) [inline packages/nextjs/next.config.js:1]: There should be no changes made to this file. Shouldn't be touched since he only focus on the builders overview issue https://github.com/BuidlGuidl/batch11.buidlguidl.com/issues/4.

File: packages/nextjs/app/page.tsx [hunk 4]
```diff
@@ -2,48 +2,106 @@
...
+  const displayCount = isLoading ? (
+    <span className="loading loading-spinner loading-sm text-primary"></span>
+  ) : error ? (
+    "Error loading"
+  ) : typeof checkedInCount === "bigint" ? (
+    checkedInCount.toString()
+  ) : (
+    "0"
+  );
+
```
Comment (phipsae) [inline packages/nextjs/app/page.tsx:28]: This is out of scope and relates to issue #6. Please remove it.

File: packages/nextjs/app/page.tsx [hunk 5]
```diff
@@ -2,48 +2,106 @@
...
+              <p className="text-base-content dark:text-white font-medium">
+                Start your journey by completing the{" "}
+                <Link
+                  href="/challenges"
+                  className="link link-primary dark:text-white font-semibold hover:link-primary transition-colors"
+                >
+                  Batch Challenges
+                </Link>
+              </p>
+            </div>
```
Comment (phipsae) [inline packages/nextjs/app/page.tsx:99]: Out of scope, that is for another issue.

File: packages/nextjs/app/page.tsx [hunk 6]
```diff
@@ -2,48 +2,106 @@
...
+          <p className="text-xl text-base-content dark:text-white font-medium mb-6">
+            Get started by taking a look at your batch GitHub repository and join our community of builders.
           </p>
+          <div className="inline-block bg-base-200 dark:bg-base-300 rounded-xl p-4 shadow-md">
+            <p className="text-lg flex gap-3 items-center justify-center">
+              <span className="font-bold text-base-content dark:text-white">Checked in builders:</span>
+              <span className="text-primary dark:text-primary-content font-mono font-bold">{displayCount}</span>
+            </p>
+          </div>
         </div>
```
Comment (phipsae) [inline packages/nextjs/app/page.tsx:46]: Out of scope, that is for another issue.

File: packages/nextjs/app/page.tsx [hunk 7]
```diff
@@ -2,48 +2,106 @@
...
-              <p>
+            <div className="flex flex-col bg-base-100 dark:bg-base-300 px-8 py-8 text-center items-center max-w-xs rounded-2xl shadow-lg hover:shadow-xl transition-all duration-200 border border-base-200 dark:border-base-400">
+              <MagnifyingGlassIcon className="h-10 w-10 text-primary dark:text-primary-content mb-4" />
+              <p className="text-base-content dark:text-white font-medium">
                 Explore your local transactions with the{" "}
-                <Link href="/blockexplorer" passHref className="link">
+                <Link
+                  href="/blockexplorer"
+                  className="link link-primary dark:text-white font-semibold hover:link-primary transition-colors"
+                >
```
Comment (phipsae) [inline packages/nextjs/app/page.tsx:70]: Out of scope, that is for another issue.

File: packages/nextjs/components/batch/BuildersGrid.tsx
Comment (phipsae) [inline packages/nextjs/components/batch/BuildersGrid.tsx:1]: As @melanke suggested, it would be great to display all the builders who have checked in, along with those who already have their personal pages.

**Review by phipsae — CHANGES_REQUESTED** [review]
Thanks for the changes! 

Here’s a general tip: focus solely on the issue you’ve chosen to work on and avoid making unrelated changes. Only modify the files necessary for that specific task.

I understand it can be challenging at first, but that’s exactly why we have this course to guide you! 😊

File: packages/nextjs/components/batch/BuildersGrid.tsx (hunk 3 above)
Comment (phipsae) [reply packages/nextjs/components/batch/BuildersGrid.tsx]: @gboigwe, this one is still open. We want to display all checked-in builders, not just those with a personal profile.

File: packages/nextjs/app/page.tsx (hunk 4 above)
Comment (phipsae) [reply packages/nextjs/app/page.tsx]: Please remove this. It belongs to issue #6. If you want to include the "checkedInCounter", create a separate PR for it.

In a "real-world" scenario, when working on an issue, ensure you only add code that falls within its scope.

File: packages/nextjs/components/batch/BuildersGrid.tsx
Comment (phipsae) [reply packages/nextjs/components/batch/BuildersGrid.tsx:1]: @gboigwe, this one is still open. We want to display all checked-in builders, not just those with a personal profile.

**Comment by phipsae** [comment]
Hi @gboigwe, thanks for the updates!

There are just two more points left to address. Please always make sure to review the conversation for details.

1) Instead of limiting to builders with profile pages, we sould include everyone who has checked in, even those without a profile. Refer to this [discussion](https://github.com/BuidlGuidl/batch11.buidlguidl.com/pull/20#discussion_r1861621550). You can use the CheckedIn event: ` event CheckedIn(bool first, address builder, address checkInContract)`. 
2) Pls don't solve other issue in this PR [discussion](https://github.com/BuidlGuidl/batch11.buidlguidl.com/pull/20#discussion_r1861624428) (no checkedIn counter, that's another issue, if you want to solve it, create another PR)

Almost done :)

**Comment by gboigwe** [comment]
Okay, thank you. Thanks for you patience with me too

**Comment by gboigwe** [comment]
I will effect that

**Comment by phipsae** [comment]
> Okay, thank you. Thanks for you patience with me too

Of course, that's why I am here! So no worries at all. It's all about learning.
@melanke thank you so much!

**Comment by melanke** [comment]
@phipsae This code doesn't work because the `fromBlock`:
```ts
const { data: events, isLoading: eventsLoading } = useScaffoldEventHistory({
    contractName: "BatchRegistry",
    eventName: "CheckedIn",
    fromBlock: 127324219n,
  });
  ```
  Which is strange, because I got the Contract Creation block and subtracted 10, as we did on the challenges:
  
image

But it works when we put zero on fromBlock.

Do you know what's happening?

**Comment by phipsae** [comment]
> @phipsae This code doesn't work because the `fromBlock`:
> 
> ```ts
> const { data: events, isLoading: eventsLoading } = useScaffoldEventHistory({
>     contractName: "BatchRegistry",
>     eventName: "CheckedIn",
>     fromBlock: 127324219n,
>   });
> ```
> 
> Which is strange, because I got the Contract Creation block and subtracted 10, as we did on the challenges:
> 
> image
> 
> But it works when we put zero on fromBlock.
> 
> Do you know what's happening?

No idea, right now. As soon as you push the code, I can have a look at it.

**Comment by phipsae** [comment]
> > @phipsae This code doesn't work because the `fromBlock`:
> > ```ts
> > const { data: events, isLoading: eventsLoading } = useScaffoldEventHistory({
> >     contractName: "BatchRegistry",
> >     eventName: "CheckedIn",
> >     fromBlock: 127324219n,
> >   });
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Which is strange, because I got the Contract Creation block and subtracted 10, as we did on the challenges:
> > image
> > But it works when we put zero on fromBlock.
> > Do you know what's happening?
> 
> No idea, right now. As soon as you push the code, I can have a look at it.

I tried it out and works on my local environment.

File: packages/nextjs/components/batch/BuildersGrid.tsx [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+      <div className="w-full px-6 py-8 text-center">
+        <span className="loading loading-spinner loading-lg text-primary"></span>
+      </div>
+    );
+  }
+
+  const buildersWithoutProfile = (events ?? [])
+    .map(event => event.args.builder)
+    .filter(address => address && !profileBuilders.includes(address));
+  console.log(buildersWithoutProfile);
```
Comment (phipsae) [inline packages/nextjs/components/batch/BuildersGrid.tsx:35]: pls remove

File: packages/nextjs/components/batch/BuildersGrid.tsx [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+            href={`/builders/${address}`}
+            className="bg-base-100 rounded-xl p-6 shadow-lg hover:shadow-xl transition-all duration-200 border border-base-200"
+          >
+            <div className="flex flex-col gap-2">
+              <Address address={address} size="lg" />
+              <div className="flex justify-between items-center">
+                <span className="text-sm font-semibold hover:underline">View Profile →</span>
+              </div>
+            </div>
+          </Link>
```
Comment (phipsae) [inline packages/nextjs/components/batch/BuildersGrid.tsx:52]: Please fix the error.
As a rule of thumb, always check the console for any errors before submitting your code.

Screenshot 2024-12-07 at 11 52 43

**Review by phipsae — CHANGES_REQUESTED** [review]
Looks great, thanks! 
Only these little comments below and then we are finally ready to merge.

File: packages/nextjs/components/batch/BuildersGrid.tsx (hunk 8 above)
Comment (melanke) [reply packages/nextjs/components/batch/BuildersGrid.tsx:35]: Oops

**Comment by phipsae** [comment]
Thanks a lot for your effort! Finally lets merge 🎉
